File: chapter_02_排序/3.insertion_sort.go
package main

//import "fmt"

/*
   插入排序

   从i=1 开始，将元素插入已经排定的数组中。
   [0, i) 是已经排定的元素
   每次循环的目的，是将 i 插入 [0, i) 中
 */
func InsertionSort(data Interface) {
	for i:=1; i<data.Len(); i++ {
		for j:=i; j>0 && data.Less(j, j-1); j-- {
			data.Swap(j, j-1)
		}
	}
}

/*
   希尔排序  插入排序的改进版本，快的一笔
 */
func ShellSort(data Interface) {
	var (
		l, h = data.Len(), 1
	)

	for {
		if h >= l / 3 {
			break
		}
		h = h * 3 + 1
	}

	for {

		if h < 1 {
			break
		}
		// hi数组变为 h有序
		for i:=h; i<l; i++ {
			for j:=i; j>=h && data.Less(j, j-h); j-=h {
				data.Swap(j, j-h)
			}
		}
		h = h / 3
	}
}


//func main() {
//	p := generateIntSlice(100000, 10)
//	fmt.Println("插入排序前")
//	//InsertionSort(p)
//	//fmt.Println("插入排序后：", p.IsSorted())
//
//	p.Shuffle()
//	ShellSort(p)
//	fmt.Println("希尔排序后：", p.IsSorted())
//}

File: chapter_02_排序/py/3_insertion_sort.py
# coding: utf-8

"""
插入排序 希尔排序(h-sort)
"""

import sys
import time
from sort_base import DataInterface, IntData, generateIntData, time_it


class InsertionSort(object):
    """将元素插入已经排定的列表里边"""
    @staticmethod
    @time_it
    def insertion_sort(data):
        for i in xrange(1, data.len()):
            for j in xrange(i, 0, -1):
                if not data.less(j, j-1):
                    break
                data.swap(j, j-1)


class ShellSort(object):
    """希尔排序：插入排序的改进版本

    """
    @staticmethod
    @time_it
    def shell_sort(data):
        # 1/ find h
        h = 1
        l = data.len()
        while h <= l / 3:
            h = 3 * h + 1

        # 2/ desc h
        while h >= 1:
            for i in xrange(1, l-h+1):
                for j in xrange(i, h-1, -h): # 每次往前步进 h
                    if not data.less(j, j-h):
                        break
                    data.swap(j, j-h)

            h /= 3


if __name__ == '__main__':

    d = generateIntData(10000, 10)

    print d.is_sorted()

    InsertionSort.insertion_sort(d)
    print d.is_sorted()

    d.shuffle()
    ShellSort.shell_sort(d)
    print d.is_sorted()